File: src/Sketch.Shared/Services/Subscriptions/SubscriptionService.cs
using System;
using System.Net.Http;
using System.Net.Http.Json;
using System.Threading.Tasks;

using Microsoft.AspNetCore.SignalR;
using Microsoft.AspNetCore.SignalR.Client;

using Sketch.Shared.Data;
using Sketch.Shared.Services;

namespace Sketch.Shared.Services
{
    public class SubscriptionService : ISubscriptionService
    {
        private readonly HttpClient _http;
        private readonly HubConnection _hubConnection;

        public SubscriptionService(HttpClient http)
        {
            _hubConnection = new HubConnectionBuilder()
                .WithUrl(new Uri(http.BaseAddress, "/hub"))
                .Build();

            _http = http;
            _hubConnection.StartAsync();
        }

        public string SubscriberId
        {
            get { return _hubConnection.ConnectionId; }
        }

        public async Task RegisterAsync(User user)
        {
            await _http.PostAsJsonAsync($"/api/identity/{SubscriberId}", user);
        }

        public async Task SubscribeAsync(string channel)
        {
            await _http.GetAsync($"/api/subscribe/{SubscriberId}/{channel}");
        }

        public async Task UnsubscribeAsync(string channel)
        {
            await _http.GetAsync($"/api/unsubscribe/{SubscriberId}/{channel}");
        }

        public IDisposable OnReceive<T>(Func<T, Task> handler)
        {
            return _hubConnection.On($"{typeof(T)}", handler);
        }

        public Task SendAsync<T>(string methodGroup, T content)
        {
            return _http.PostAsJsonAsync($"/api/{methodGroup}/{SubscriberId}", content);
        }

        public Task SendAsync<T>(string methodGroup, string methodName, T content)
        {
            return _http.PostAsJsonAsync($"/api/{methodGroup}/{methodName}/{SubscriberId}", content);
        }
    }
}

File: src/Sketch.WebApp.JS/src/services/SubscriptionService.js
import { HubConnectionBuilder, HubConnectionState } from "@microsoft/signalr";

const ApplicationBaseAddress = "https://localhost:5001";

class SubscriptionService {

  constructor() {

    this.connection = new HubConnectionBuilder()
      .withUrl(`${ApplicationBaseAddress}/hub`)
      .withAutomaticReconnect()
      .build();
  }

  async subscribe(channel) {
    if (this.connection.state !== HubConnectionState.Connected) {
      await this.connection.start();
    }

    await fetch(`${ApplicationBaseAddress}/api/subscribe/${this.connection.connectionId}/${channel}`);
  }

  async unsubscribe(channel) {
    if (this.connection.state !== HubConnectionState.Connected) {
      await this.connection.start();
    }

    await fetch(`${ApplicationBaseAddress}/api/unsubscribe/${this.connection.connectionId}/${channel}`);
  }

  async send(methodGroup, data) {
    return await fetch(`${ApplicationBaseAddress}/api/${methodGroup}/${this.connection.connectionId}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(data)
    })
  }

  async sendToMethod(methodGroup, methodName, data) {
    return await fetch(`${ApplicationBaseAddress}/api/${methodGroup}/${methodName}/${this.connection.connectionId}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(data)
    });
  }

  on(eventType, eventHandler) {
    console.log(`on - Sketch.Shared.Data.${eventType}`);
    this.connection.on(`Sketch.Shared.Data.${eventType}`, eventHandler);
  }

  off(eventType, eventHandler) {
    console.log(`off - Sketch.Shared.Data.${eventType}`);
    this.connection.off(`Sketch.Shared.Data.${eventType}`, eventHandler);
  }
}

export const Subscription = new SubscriptionService();
